src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/LoginHandler.cs

```diff
@@ -0,0 +1,636 @@
+                {
+                    libraryType = TdsEnums.FedAuthLibrary.SecurityToken,
+                    fedAuthRequiredPreLoginResponse = context.FedAuthNegotiatedInPrelogin,
+                    accessToken = context.AccessTokenInBytes
+                };
+                // No need any further info from the server for token based authentication. So set _federatedAuthenticationRequested to true
+                features.FederatedAuthenticationRequested = true;
+            }
+
+            // The GLOBALTRANSACTIONS, DATACLASSIFICATION, TCE, and UTF8 support features are implicitly requested
+            requestedFeatures |= TdsEnums.FeatureExtension.GlobalTransactions | TdsEnums.FeatureExtension.DataClassification | TdsEnums.FeatureExtension.Tce | TdsEnums.FeatureExtension.UTF8Support;
+
+            // The SQLDNSCaching feature is implicitly set
+            requestedFeatures |= TdsEnums.FeatureExtension.SQLDNSCaching;
+
+            features.RequestedFeatures = requestedFeatures;
+
+            await TdsLogin(context, isAsync, ct).ConfigureAwait(false);
+
+            // If the workflow being used is Active Directory Authentication and server's prelogin response
+            // for FEDAUTHREQUIRED option indicates Federated Authentication is required, we have to insert FedAuth Feature Extension
+            // in Login7, indicating the intent to use Active Directory Authentication for SQL Server.
+            static bool ShouldRequestFedAuth(LoginHandlerContext context)
+            {
+                bool IsEntraIdAuthInConnectionString = context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryPassword
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryInteractive
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryDeviceCodeFlow
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryServicePrincipal
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryManagedIdentity
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryMSI
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryDefault
+                                                || context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryWorkloadIdentity
+                                                // Since AD Integrated may be acting like Windows integrated, additionally check _fedAuthRequired
+                                                || (context.ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryIntegrated);
```

Comment:
You should be including a check for _fedAuthRequired with AD Integrated?

Reply:
This is the check we do today while login. 
I didn't quite understand the suggestion. Are you asking whether we should continue checking for AD integrated? Or are you saying that we could remove it ?

Reply:
This is how it's compared in old code:

(ConnectionOptions.Authentication == SqlAuthenticationMethod.ActiveDirectoryIntegrated && _fedAuthRequired)

Reply:
1. I should have deleted that comment.
2. The _fedAuthRequired has been renamed to `FedAuthNegotiatedInPrelogin` in the context object and is being honored in the next line of code.. The _fedAuthRequired was set in ConsumePrelogin in SqlInternalConnectionTds. In this code, the PReloginhandler sets `FedAuthNegotiatedInPrelogin` with proper naming assigned to it. 

This function faithfully represents what was being done earlier with a slight change of checking for access token based on your PR comments about putting the check for AcceessTokens inside the FedAuth feature request check. 


The check for `Authentication` options has been first stored in `IsEntraIdAuthInConnectionString` to say that we are first evaluating the conditions in Connection string, after this we evaluate the conditions with Prelogin handshake as well as the other checks like callback and token bytes.